### ISSUE TITLE
Drawing Lines is Messed Up

### ISSUE THREAD

**Issue by sunjay** [ggez/ggez#137]
The following minimal example was created by taking the example from ggez.rs and then replacing the code to draw a circle with code to draw a single line. The problem still appears when more than one line is drawn. This is just a minimal example so I left it at one line.

```rust
extern crate ggez;
use ggez::conf;
use ggez::event;
use ggez::{GameResult, Context};
use ggez::graphics;
use ggez::graphics::{Color, DrawMode, Point};
use std::time::Duration;

struct MainState {
}

impl MainState {
    fn new(ctx: &mut Context) -> GameResult<MainState> {
        let s = MainState {};
        Ok(s)
    }
}

impl event::EventHandler for MainState {
    fn update(&mut self, _ctx: &mut Context, _dt: Duration) -> GameResult<()> {
        Ok(())
    }

    fn draw(&mut self, ctx: &mut Context) -> GameResult<()> {
        graphics::clear(ctx);
        // THIS IS THE ONLY IMPORTANT LINE:
        graphics::line(ctx, &[Point {x: 200., y: 200.}, Point {x: 400., y: 500.}])?;
        graphics::present(ctx);
        Ok(())
    }
}

pub fn main() {
    let c = conf::Conf::new();
    let ctx = &mut Context::load_from_conf("super_simple", "ggez", c).unwrap();
    let state = &mut MainState::new(ctx).unwrap();
    event::run(ctx, state).unwrap();
}
```

## Screenshot:

![screenshot from 2017-09-02 00-17-53](https://user-images.githubusercontent.com/530939/29993440-38a2c810-8f74-11e7-8c03-8e2dfc9fbf98.png)

## Expected:

A simple line from (200, 200) to (400, 500)

## Actual:

A really skewed line that stretches way past those points and looks really messed up.

## More Information

This seems to only happen with irregular slopes (!= 1.0). If I draw a line between (200, 200) and (400, 400), it works. Also, if I add another point to the line like (100, 100) or something, it suddenly works again.

With the draw call changed to:
```rust
        graphics::line(ctx, &[Point {x: 200., y: 200.}, Point {x: 400., y: 500.}, Point {x: 100., y: 100.}])?;
```

Results in:
![screenshot from 2017-09-02 00-34-39](https://user-images.githubusercontent.com/530939/29993524-8793024e-8f76-11e7-9c3c-a678e10fac67.png)

However, the following:
```rust
        graphics::line(ctx, &[Point {x: 200., y: 200.}, Point {x: 400., y: 500.}, Point {x: 600., y: 400.}])?;
```

Results in:
![screenshot from 2017-09-02 00-35-45](https://user-images.githubusercontent.com/530939/29993526-b5dd410a-8f76-11e7-8575-fdf71d85d8c5.png)

## Platform

<details>
    <summary>System Information</summary>
<pre>
System:    Host: brown-dynamite Kernel: 4.10.0-32-generic x86_64 (64 bit gcc: 5.4.0)
           Desktop: Cinnamon 3.4.6 (Gtk 3.18.9) dm: mdm Distro: Linux Mint 18.2 Sonya
Machine:   System: ASUSTeK (portable) product: X556URK v: 1.0
           Mobo: ASUSTeK model: X556URK v: 1.0 Bios: American Megatrends v: X556URK.304 date: 01/19/2017
CPU:       Dual core Intel Core i5-7200U (-HT-MCP-) cache: 3072 KB
           flags: (lm nx sse sse2 sse3 sse4_1 sse4_2 ssse3 vmx) bmips: 10848
           clock speeds: min/max: 400/3100 MHz 1: 871 MHz 2: 871 MHz 3: 874 MHz 4: 821 MHz
Graphics:  Card-1: Intel Device 5916 bus-ID: 00:02.0 chip-ID: 8086:5916
           Card-2: NVIDIA Device 134e bus-ID: 01:00.0 chip-ID: 10de:134e
           Display Server: X.Org 1.18.4 drivers: intel (unloaded: fbdev,vesa) FAILED: nouveau
           Resolution: 1920x1080@60.05hz
           GLX Renderer: Mesa DRI Intel HD Graphics 620 (Kabylake GT2)
           GLX Version: 3.0 Mesa 17.0.7 Direct Rendering: Yes
Audio:     Card Intel Device 9d71 driver: snd_hda_intel bus-ID: 00:1f.3 chip-ID: 8086:9d71
           Sound: Advanced Linux Sound Architecture v: k4.10.0-32-generic
Network:   Card-1: Realtek RTL8111/8168/8411 PCI Express Gigabit Ethernet Controller
           driver: r8169 v: 2.3LK-NAPI port: d000 bus-ID: 02:00.0 chip-ID: 10ec:8168
           IF: enp2s0 state: down mac: <filter>
           Card-2: Qualcomm Atheros Device 0042 driver: ath10k_pci bus-ID: 03:00.0 chip-ID: 168c:0042
           IF: wlp3s0 state: up speed: N/A duplex: N/A mac: <filter>
Drives:    HDD Total Size: 1000.2GB (41.0% used)
           ID-1: /dev/sda model: TOSHIBA_MQ01ABD1 size: 1000.2GB serial: 176NT4SAT
Partition: ID-1: / size: 75G used: 27G (37%) fs: ext4 dev: /dev/sda7
           ID-2: /home size: 737G used: 345G (50%) fs: ext4 dev: /dev/sda8
           ID-3: swap-1 size: 12.76GB used: 0.00GB (0%) fs: swap dev: /dev/sda6
RAID:      System: supported: N/A
           No RAID devices: /proc/mdstat, md_mod kernel module present
           Unused Devices: none
Sensors:   System Temperatures: cpu: 41.5C mobo: N/A
           Fan Speeds (in rpm): cpu: N/A
Repos:     Active apt sources in file: /etc/apt/sources.list.d/dotnetdev.list
           deb [arch=amd64] https: //apt-mo.trafficmanager.net/repos/dotnet-release/ xenial main
           Active apt sources in file: /etc/apt/sources.list.d/google-chrome.list
           deb [arch=amd64] http: //dl.google.com/linux/chrome/deb/ stable main
           Active apt sources in file: /etc/apt/sources.list.d/hvr-ghc-xenial.list
           deb http: //ppa.launchpad.net/hvr/ghc/ubuntu xenial main
           deb-src http: //ppa.launchpad.net/hvr/ghc/ubuntu xenial main
           Active apt sources in file: /etc/apt/sources.list.d/jonathonf-vim-xenial.list
           deb http: //ppa.launchpad.net/jonathonf/vim/ubuntu xenial main
           deb-src http: //ppa.launchpad.net/jonathonf/vim/ubuntu xenial main
           Active apt sources in file: /etc/apt/sources.list.d/nathan-renniewaldock-flux-xenial.list
           deb http: //ppa.launchpad.net/nathan-renniewaldock/flux/ubuntu xenial main
           deb-src http: //ppa.launchpad.net/nathan-renniewaldock/flux/ubuntu xenial main
           Active apt sources in file: /etc/apt/sources.list.d/nodesource.list
           deb https: //deb.nodesource.com/node_8.x xenial main
           deb-src https: //deb.nodesource.com/node_8.x xenial main
           Active apt sources in file: /etc/apt/sources.list.d/official-package-repositories.list
           deb http: //mirror.csclub.uwaterloo.ca/linuxmint-packages sonya main upstream import backport
           deb http: //archive.ubuntu.com/ubuntu xenial main restricted universe multiverse
           deb http: //archive.ubuntu.com/ubuntu xenial-updates main restricted universe multiverse
           deb http: //archive.ubuntu.com/ubuntu xenial-backports main restricted universe multiverse
           deb http: //security.ubuntu.com/ubuntu/ xenial-security main restricted universe multiverse
           deb http: //archive.canonical.com/ubuntu/ xenial partner
           Active apt sources in file: /etc/apt/sources.list.d/peek-developers-stable-xenial.list
           deb http: //ppa.launchpad.net/peek-developers/stable/ubuntu xenial main
           deb-src http: //ppa.launchpad.net/peek-developers/stable/ubuntu xenial main
           Active apt sources in file: /etc/apt/sources.list.d/saiarcot895-myppa-xenial.list
           deb http: //ppa.launchpad.net/saiarcot895/myppa/ubuntu xenial main
           deb-src http: //ppa.launchpad.net/saiarcot895/myppa/ubuntu xenial main
           Active apt sources in file: /etc/apt/sources.list.d/slack.list
           deb https: //packagecloud.io/slacktechnologies/slack/debian/ jessie main
           Active apt sources in file: /etc/apt/sources.list.d/swi-prolog-stable-xenial.list
           deb http: //ppa.launchpad.net/swi-prolog/stable/ubuntu xenial main
           deb-src http: //ppa.launchpad.net/swi-prolog/stable/ubuntu xenial main
           Active apt sources in file: /etc/apt/sources.list.d/ubuntuhandbook1-sigil-xenial.list
           deb http: //ppa.launchpad.net/ubuntuhandbook1/sigil/ubuntu xenial main
           deb-src http: //ppa.launchpad.net/ubuntuhandbook1/sigil/ubuntu xenial main
           Active apt sources in file: /etc/apt/sources.list.d/vscode.list
           deb [arch=amd64] http: //packages.microsoft.com/repos/vscode stable main
           Active apt sources in file: /etc/apt/sources.list.d/yarn.list
           deb https: //dl.yarnpkg.com/debian/ stable main
Info:      Processes: 292 Uptime: 14 days Memory: 4916.0/11894.6MB
           Init: systemd v: 229 runlevel: 5 default: 2 Gcc sys: 5.4.0
           Client: Unknown python2.7 client inxi: 2.2.35
</pre>
</details>

**Comment by termhn** [comment]
I can't reproduce this on my machine... what version of ggez are you using?

![image](https://user-images.githubusercontent.com/887711/29994732-355e087e-8f8b-11e7-8c78-19d5f3cba98c.png)

**Comment by termhn** [comment]
Can you clone this: https://github.com/termhn/ggez-goodies

and run 

```sh
cargo run --example camera_test --release
```

I use lines fairly extensively in that example.

**Comment by sunjay** [comment]
```
ggez 0.3.3
```

Should be the latest one since I just started using this yesterday. Could it be my platform? I'm on Linux Mint. Full platform info is in my original issue at the bottom.

I tried the example you mentioned and it renders fine. I checked and it looks like you're using master right now. I switched my code to use master and it renders fine now.

So while this is definitely still a bug in the latest release, it looks like this is fixed in master. Probably good to close this issue once whatever is on master is published to crates.io. :smile: 

I'll use master for now.

**Comment by termhn** [comment]
Interesting. Yeah, I can repro on 0.3.3 as well. 

This seems to be the change https://github.com/ggez/ggez/compare/0.3...master#diff-3d38872253ac1a673a47a51a3a5c5517R1133

**Comment by sunjay** [comment]
Must be because of the upgraded dependencies. Would be great if a new version could be published ASAP so no one else runs into this.

**Comment by icefoxen** [comment]
Is this at all related to https://github.com/ggez/ggez/issues/107 ?

**Comment by icefoxen** [comment]
Yeah now that I double check there's no really good way to switch to new Lyon without an API break. So we can't really make an 0.3.4 with those changes, it has to be 0.4.0.  

I always get annoyed when known bugs get marked "won't fix" but that's sort of the only solution I have right now...

**Comment by sunjay** [comment]
Is upgrading to that version a big deal? I'm fine with this being on the latest version of the library

**Comment by icefoxen** [comment]
It would break backwards compatibility; the public-facing API would change from 0.3.3 to 0.3.4.  I guess the semver requirements technically allow it since we're pre-1.0 but I'd really rather not do it.

**Comment by sunjay** [comment]
I meant that I wouldn't mind upgrading from 0.3 to 0.4. I totally understand if it's hard to backport the changes. Does the library change a lot between 0.3 and 0.4? Do you think people won't want to upgrade?

**Comment by icefoxen** [comment]
Oh!  I misunderstood.  All right then.  :-)

0.4 is currently a work-in-progress, so you can follow the git master if you want; I try to keep it generally usable.  No *huge* breaking changes are planned but it looks like there may be a number of small ones; the Lyon upgrade is one of those.  It's currently mostly done and seems to be working well, see issue #116 for details.

**Comment by sunjay** [comment]
That works! Glad to see that this will be fixed in 0.4. Might be worth putting a link to this issue in the 0.3 README or something if more people end up effected. :)

**Comment by icefoxen** [comment]
All right then, thank you.

If you want to develop against the git master for now I try to keep it in working condition.
